File: C#/Socketry/socket/SocketDebug.cs
using System;
using System.Collections.Generic;
using System.IO.Pipes;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;

namespace socket
{
    public class SocketDebug : ISocket
    {
        PipeStream readPipe, writePipe;
        bool connected;
        bool blocking;

        public SocketDebug()
        {
            readPipe = new AnonymousPipeServerStream();
            writePipe = new AnonymousPipeClientStream("writePipe");
        }

        public bool IsConnected()
        {
            return connected;
        }

        public bool IsBlocking()
        {
            return blocking;
        }

        public int Read(MemoryStream dst)
        {
            int r =readPipe.Read(dst.GetBuffer());
            return r;
        }

        public int Write(MemoryStream src)
        {
            writePipe.Write(src.GetBuffer());
            return 0;
        }

        public Socket ConfigureBlocking(bool blocking)
        {
            this.blocking = blocking;
            return new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
        }

        public bool Connect(EndPoint remote)
        {
            connected = true;
            return true;
        }

        public SelectionKey Register(Selector sel, int ops, Object att)
        {
            //return sel.register(sel,ops,att);
            return null;
        }

        public int ReadOtherSide(MemoryStream dst)
        {
            int r = writePipe.Read(dst.GetBuffer());
            return r;
        }

        public int WriteOtherSide(MemoryStream src)
        {
            readPipe.Write(src.GetBuffer());
            return 0;
        }
    }
}

File: JAVA/socketry-main/src/main/java/com/socketry/socket/SocketDebug.java
package com.socketry.socket;

import java.io.IOException;
import java.net.SocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.ClosedChannelException;
import java.nio.channels.Pipe;
import java.nio.channels.SelectableChannel;
import java.nio.channels.SelectionKey;
import java.nio.channels.Selector;



public class SocketDebug implements ISocket {
    Pipe readPipe, writePipe;
    boolean connected;
    boolean blocking;

    public SocketDebug() throws IOException {
        readPipe = Pipe.open();
        writePipe = Pipe.open();

        connected = false;
    }

    @Override
    public int read(ByteBuffer dst) throws IOException {
        int r = readPipe.source().read(dst);
        return r;
    }

    @Override
    public int write(ByteBuffer src) throws IOException {
        return writePipe.sink().write(src);
    }

    @Override
    public SelectableChannel configureBlocking(boolean block) throws IOException {
        readPipe.source().configureBlocking(block);
        readPipe.sink().configureBlocking(block);
        writePipe.source().configureBlocking(block);
        writePipe.sink().configureBlocking(block);
        blocking = block;
        return readPipe.source();
    }

    @Override
    public boolean connect(SocketAddress remote) throws IOException {
        connected = true;
        return true;
    }

    @Override
    public boolean isConnected() {
        return connected;
    }

    @Override
    public boolean isBlocking() {
        return blocking;
    }

    @Override
    public SelectionKey register(Selector sel, int ops, Object att) throws ClosedChannelException {
        return readPipe.source().register(sel, ops, att);
    }

    public int readOtherSide(ByteBuffer dst) throws IOException {
        return writePipe.source().read(dst);
    }

    public int writeOtherSide(ByteBuffer src) throws IOException {
        System.out.println("Writing to other side : " + src.array().length);
        return readPipe.sink().write(src);
    }
}